lib/semantics/resolve-names.cc

```diff
@@ -37,31 +37,41 @@ class MessageHandler;
 
 static GenericSpec MapGenericSpec(const parser::GenericSpec &);
 
-// ImplicitRules maps initial character of identifier to the DeclTypeSpec*
-// representing the implicit type; nullptr if none.
+// ImplicitRules maps initial character of identifier to the DeclTypeSpec
+// representing the implicit type; std::nullopt if none.
+// It also records the presence of IMPLICIT NONE statements.
+// When inheritFromParent is set, defaults come from the parent rules.
 class ImplicitRules {
 public:
-  ImplicitRules(MessageHandler &messages);
-  bool isImplicitNoneType() const { return isImplicitNoneType_; }
-  bool isImplicitNoneExternal() const { return isImplicitNoneExternal_; }
+  ImplicitRules(MessageHandler &messages)
+    : messages_{messages}, inheritFromParent_{false} {}
+  ImplicitRules(std::unique_ptr<ImplicitRules> &&parent)
+    : messages_{parent->messages_}, inheritFromParent_{true} {
+    parent_.swap(parent);
+  }
+  std::unique_ptr<ImplicitRules> &&parent() { return std::move(parent_); }
```

Comment:
I have never before seen a function return an rvalue reference.  I wonder whether there are other interfaces in f18 code that return lvalue references that should be rvalues?  (No action req'd)